strapi/strapi
# test: add e2e tests for CTB restarts from file changes

### What does it do?

- adds e2e test for creating a collection type and that the ctb tutorial modal appears
- adds util to watch for the server restart
- adds util to help ignore the ctb tutorial modal
- adds a very hacky test framework and util for resetting the filesystem state after changes like CTB edits, uploads, etc
- minor refactoring and docs

I'm not super happy with how hacky it is, but it seems stable enough from the time I've been working with it. The most important thing I would like to find a solution for is the `delay()` to wait for the server to realize files have changed and it begins its restart process. However, without access to the output of the server I can't think of anything better. Perhaps a custom winston logger could pipe it somewhere that we could watch that would be realtime?

### Why is it needed?

We were limited by being unable to test the CTB because of the server restarts, filesystem changes, etc.

### Roadmap

After this PR, I am planning the following changes:

- bring CLI tests from `v5/main` to `develop` (and then work from there on further changes)
- copy/paste and clean up first attempt at CTB tests from https://github.com/strapi/strapi/pull/18065
- refactor e2e+cli tests to share code, ideally everything except the test runner itself (jest vs playwright command)
- lots of other little cleanups, like improving the restart monitoring being added here

### How to test it?

Running the full test suite with multiple browsers (that is, re-using the same running instance of strapi across multiple test runs of the same suite, and re-using the test-apps between runs ie without using `--setup`) should work every time. And continuing to run them many times in a row, they should always succeed without flakiness.

### Related issue(s)/PR(s)

DX-1166
